tests/func-tests/node_placement_test.go

```diff
@@ -17,39 +24,112 @@ import (
 
 	networkaddonsv1 "github.com/kubevirt/cluster-network-addons-operator/pkg/apis/networkaddonsoperator/v1"
-	"kubevirt.io/client-go/kubecli"
-	"kubevirt.io/kubevirt/tests/flags"
-
 	tests "github.com/kubevirt/hyperconverged-cluster-operator/tests/func-tests"
+	"kubevirt.io/client-go/kubecli"
 )
 
-const nodePlacementSamplingSize = 20
-
-var _ = Describe("[rfe_id:4356][crit:medium][vendor:cnv-qe@redhat.com][level:system]Node Placement", func() {
+const (
+	hcoLabel    = "node.kubernetes.io/hco-test-node-type"
+	infra       = "infra"
+	workloads   = "workloads"
+	hcoResource = "hyperconvergeds"
+)
 
+var _ = Describe("[rfe_id:4356][crit:medium][vendor:cnv-qe@redhat.com][level:system]Node Placement", Ordered, func() {
 	var workloadsNode *v1.Node
-
+	ctx := context.TODO()
 	tests.FlagParse()
 	client, err := kubecli.GetKubevirtClient()
 	kvtutil.PanicOnError(err)
 
-	workloadsNodes, err := client.CoreV1().Nodes().List(context.TODO(), k8smetav1.ListOptions{
-		LabelSelector: "node.kubernetes.io/hco-test-node-type==workloads",
+	BeforeAll(func() {
+		nodes, err := client.CoreV1().Nodes().List(context.TODO(), k8smetav1.ListOptions{LabelSelector: "node-role.kubernetes.io/control-plane!="})
+		kvtutil.PanicOnError(err)
+		if len(nodes.Items) < 2 {
+			Skip("Skipping Node Placement tests due to insufficient cluster nodes")
+		}
+		// Label all but first node with "node.kubernetes.io/hco-test-node-type=infra"
+		// We are doing this to remove dependency of this Describe block on a shell script that
+		// labels the nodes this way
+		for _, node := range nodes.Items[:len(nodes.Items)-1] {
+			err = setHcoNodeTypeLabel(client, &node, infra)
+			kvtutil.PanicOnError(err)
+		}
+		// Label the last node with "node.kubernetes.io/hco-test-node-type=workloads"
+		err = setHcoNodeTypeLabel(client, &nodes.Items[len(nodes.Items)-1], workloads)
+		kvtutil.PanicOnError(err)
+
+		// modify the HCO CR to use the labels we just applied to the nodes
+		originalHco := tests.GetHCO(ctx, client)
+		// modify the "infra" and "workloads" keys
+		infraVal := hcov1beta1.HyperConvergedConfig{
+			NodePlacement: &sdkapi.NodePlacement{
+				NodeSelector: map[string]string{hcoLabel: infra},
+			},
+		}
+		workloadsVal := hcov1beta1.HyperConvergedConfig{
+			NodePlacement: &sdkapi.NodePlacement{
+				NodeSelector: map[string]string{hcoLabel: workloads},
+			},
+		}
+
+		hco := &hcov1beta1.HyperConverged{}
+		originalHco.DeepCopyInto(hco)
+
+		hco.Spec.Infra = infraVal
+		hco.Spec.Workloads = workloadsVal
+		hco.SetGroupVersionKind(schema.GroupVersionKind{
+			Group:   util.APIVersionGroup,
+			Version: util.CurrentAPIVersion,
+			Kind:    util.HyperConvergedKind,
+		})
+		tests.UpdateHCORetry(ctx, client, hco)
+
+		workloadsNodes, err := client.CoreV1().Nodes().List(context.TODO(), k8smetav1.ListOptions{
+			LabelSelector: "node.kubernetes.io/hco-test-node-type==workloads",
+		})
+		kvtutil.PanicOnError(err)
+		if workloadsNodes != nil && len(workloadsNodes.Items) == 1 {
```

Comment:
no need to check both no-nil and length. The len() methods returns 0 for nil slices.

Reply:
Actually, I forgot to remove this check. Now that we have already added a check at the time of fetching the nodes in the beginning of `BeforeAll`, I think we can get rid of this check entirely. WDYT?